File: client/typescript/src/_impl/mir/metadata/layout.ts
import type {
    ElementLayout,
    Layout,
    LayoutField,
    LayoutId,
    LayoutMetadata,
    LayoutShape,
    LayoutTable,
} from "../../../_generated/mir/metadata/layout.js";
import type { LocalNodeId } from "../../../_generated/mir/tree/node.js";
import { getLocalNodeValue } from "../node.js";

export const LayoutMetadataImpl = {
    /** Return the layout entry for a type id when available. */
    typeLayout(metadata: LayoutMetadata, ty: LocalNodeId): Layout | undefined {
        const layout = LayoutMetadataImpl.layoutId(metadata, ty);

        return layout === undefined ? undefined : LayoutTableImpl.layout(metadata.layoutTable, layout);
    },

    /** Return the layout id for a type when present. */
    layoutId(metadata: LayoutMetadata, ty: LocalNodeId): LayoutId | undefined {
        return getLocalNodeValue(metadata.layoutByType, ty);
    },
};

export const LayoutTableImpl = {
    /** Return a layout entry for an id. */
    layout(table: LayoutTable, id: LayoutId): Layout | undefined {
        return table.layouts[LayoutIdImpl.index(id)];
    },
};

export const LayoutImpl = {
    /** Return the byte offset of a dynamic dispatch pointer. */
    dynamicDispatchOffset(layout: Layout): number | undefined {
        return layout.shape.kind === "dynamic" ? layout.alignment : undefined;
    },

    /** Return the byte width of this layout. */
    byteLen(layout: Layout): number {
        return layout.size;
    },

    /** Return one field by layout index. */
    fieldAt(layout: Layout, index: number): LayoutField | undefined {
        return LayoutShapeImpl.fields(layout.shape)[index];
    },

    /** Return the field count for field-addressable layouts. */
    fieldCount(layout: Layout): number | undefined {
        if (layout.shape.kind === "struct") {
            return layout.shape.struct.fields.length;
        }

        // tuple layouts address elements as fields
        if (layout.shape.kind === "tuple") {
            return layout.shape.tuple.elements.length;
        }

        // object layouts address instance fields
        if (layout.shape.kind === "object") {
            return layout.shape.object.fields.length;
        }

        return undefined;
    },
};

export const LayoutShapeImpl = {
    /** Return element layout when this shape stores indexed elements inline. */
    elements(shape: LayoutShape): ElementLayout | undefined {
        if (shape.kind === "array") {
            return shape.array;
        }

        // vectors share the same inline element layout form
        if (shape.kind === "vector") {
            return shape.vector;
        }

        return undefined;
    },

    /** Return field layouts for field-addressable shapes. */
    fields(shape: LayoutShape): ReadonlyArray<LayoutField> {
        if (shape.kind === "struct") {
            return shape.struct.fields;
        }

        // tuple layouts address elements as fields
        if (shape.kind === "tuple") {
            return shape.tuple.elements;
        }

        // object layouts address instance fields
        if (shape.kind === "object") {
            return shape.object.fields;
        }

        return [];
    },
};

export const LayoutIdImpl = {
    /** Return the raw layout identifier value. */
    raw(id: LayoutId): number {
        return id;
    },

    /** Return the zero-based layout table index. */
    index(id: LayoutId): number {
        return id - 1;
    },
};

File: client/python/src/destack/_impl/mir/metadata/layout.py
from __future__ import annotations

from destack._impl.model import ModelImpl


class LayoutMetadataImpl(ModelImpl):
    """Methods for layout metadata."""

    def type_layout(self, ty):
        """Return the layout entry for a type id when available."""

        layout = self.layout_id(ty)

        return None if layout is None else self.layout_table.layout(layout)

    def layout_id(self, ty):
        """Return the layout id for a type when present."""

        return self.layout_by_type.get(ty)


class LayoutTableImpl(ModelImpl):
    """Methods for layout tables."""

    def layout(self, id_):
        """Return a layout entry for an id."""

        return sequence_get(self.layouts, id_ - 1)


class LayoutImpl(ModelImpl):
    """Methods for concrete layouts."""

    def dynamic_dispatch_offset(self) -> int | None:
        """Return the byte offset of a dynamic dispatch pointer."""

        if self.shape.kind == "dynamic":
            return self.alignment

        return None

    def byte_len(self) -> int:
        """Return the byte width of this layout."""

        return self.size

    def field_at(self, index: int):
        """Return one field by layout index."""

        return sequence_get(self.shape.fields(), index)

    def field_count(self) -> int | None:
        """Return the field count for field-addressable layouts."""

        if self.shape.kind == "struct":
            return len(self.shape.struct.fields)

        # tuple layouts address elements as fields
        if self.shape.kind == "tuple":
            return len(self.shape.tuple.elements)

        # object layouts address instance fields
        if self.shape.kind == "object":
            return len(self.shape.object.fields)

        return None


class LayoutShapeImpl(ModelImpl):
    """Methods for layout shapes."""

    def elements(self):
        """Return element layout when this shape stores indexed elements inline."""

        if self.kind == "array":
            return self.array

        # vectors share the same inline element layout form
        if self.kind == "vector":
            return self.vector

        return None

    def fields(self):
        """Return field layouts for field-addressable shapes."""

        if self.kind == "struct":
            return self.struct.fields

        # tuple layouts address elements as fields
        if self.kind == "tuple":
            return self.tuple.elements

        # object layouts address instance fields
        if self.kind == "object":
            return self.object.fields

        return ()


def sequence_get(sequence, index: int):
    """Return one sequence item when the index is present."""

    if 0 <= index < len(sequence):
        return sequence[index]

    return None
